pytorch_blade/src/compiler/mlir/converters/mhlo_conversion.cpp

```diff
@@ -199,15 +319,103 @@ class ConvertToMhloImpl {
+  torch::jit::CompilationUnit unit;
+  auto fn = unit.create_function("main", graph);
+  auto op = torch_mlir::importJitFunctionAsFuncOp(wrap(&mlir_context), fn);
+  auto builder = ::mlir::OpBuilder(&mlir_context);
+  mlir_module.push_back(unwrap(op));
+
+  ::mlir::PassManager pm(
+      &mlir_context, ::mlir::OpPassManager::Nesting::Implicit);
+  if (enable_printing) {
+    pm.enableIRPrinting();
+  }
+  ::mlir::torch::Torch::TorchLoweringPipelineOptions options;
+  ::mlir::torch::Torch::createTorchFunctionToTorchBackendPipeline(pm, options);
+  if (mlir::failed(pm.run(mlir_module))) {
+    mlir_module.emitError() << "TorchFunctionToTorchBackendPipeline failed";
+    return std::make_tuple("", "", "", "");
+  }
+
+  ::mlir::PassManager pm2(
+      &mlir_context, ::mlir::OpPassManager::Nesting::Implicit);
+  if (enable_printing) {
+    pm2.enableIRPrinting();
+  }
+  ::mlir::torch::TorchConversion::createTorchBackendToMhloBackendPipeline(
+      pm2, options);
+  if (mlir::failed(pm2.run(mlir_module))) {
+    mlir_module.emitError() << "TorchBackendToMhloBackendPipeline failed";
+    return std::make_tuple("", "", "", "");
+  }
+
+  // Find the unique return op.
+  ::mlir::func::FuncOp funcOp;
+  ::mlir::WalkResult walkResult =
+      mlir_module.walk([&](::mlir::func::FuncOp op) {
+        if (funcOp)
+          return ::mlir::WalkResult::interrupt();
+        funcOp = op;
+        return ::mlir::WalkResult::advance();
+      });
+  if (walkResult.wasInterrupted()) {
+    mlir_module.emitError()
+        << "unimplemented: refining returns for function with "
+           "more than one return op";
+    return std::make_tuple("", "", "", "");
+  }
+  auto entry_attr =
+      GetFunctionAttrs(builder, const_graph->inputs(), const_graph->outputs());
+  auto funcType = funcOp.getFunctionType();
+  auto tf_func = builder.create<mlir::func::FuncOp>(
+      funcOp.getLoc(),
+      "main",
+      funcType,
+      ::llvm::ArrayRef<::mlir::NamedAttribute>{entry_attr});
+  ::mlir::BlockAndValueMapping mapper;
+  funcOp.cloneInto(tf_func, mapper);
+  funcOp.erase();
+  mlir_module.push_back(tf_func);
+
+  std::string parsable_str = GenerateMlirModuleString(false, mlir_module);
+  std::string pretty_str = GenerateMlirModuleString(true, mlir_module);
+  return std::make_tuple(
+      std::move(parsable_str),
+      std::move(pretty_str),
+      GetAttrString(GetDeviceStrs(const_graph->inputs())),
+      GetAttrString(GetDeviceStrs(const_graph->outputs())));
+}
+
 std::tuple<std::string, std::string, std::string, std::string>
 ConvertTorchScriptToMhlo(std::shared_ptr<torch::jit::Graph> graph) {
   try {
-    mlir::DialectRegistry registry;
-    RegisterDialects(registry);
-    mlir::MLIRContext mlir_context(registry);
-    mlir_context.loadAllAvailableDialects();
-    ConvertToMhloImpl impl(graph, mlir_context);
-    return impl.Run();
+    if (env::ReadBoolFromEnvVar("TORCH_DISC_USE_TORCH_MLIR", false)) {
```

Comment:
Looks like we are going to switch the conversion entirely to torch_mlir (instead of merging our existing converter with torch_mlir)? Are all the ops supported in torch_mlir a superset of what in our existing converter?

Reply:
The question related to 2 aspects of TorchMLIR and TorchMHLO:
1. From the aspects of the conversion framework: Yes, we are going to adopt torch_mlir since it has a better infrastructure than ours and more community support.
2. From the aspects of lowering to MHLO, TorchMLIR hasn't done anything about MHLO. Although MHLO was planned in their roadmap, the community only developed conversion to TOSA and LinAlg for now.

Reply:
In our roadmap, we would like to adopt torch_mlir's infrastructure to reduce the efforts of developing conversion to MHLO, and finally, merge our TorchToMHLO pass to upstream.